[T] Add hasChanged prop to HdForm to make tracking changes easier

majakomel opened the pull request:
I added a new watcher and emitted event to HdForm because we need a way to track if the form has been changed in parent component.

Our specific use case is that we want to confirm with the user if he really wants to leave an unfinished flow via modal, and for that we need to know about the changes.

kuroski reviewed (commented):
Maybe two things you could do is:
- Add a story about this property
- Create a `HdForm` test for this new property

😄

viniciuskneves commented on src/components/form/HdForm.vue:
```diff
@@ -24,15 +24,32 @@ export default {
       default: false,
     },
+    hasChanged: {
+      type: Boolean,
+      default: null,
```
If it is `Boolean`, I would set the default value to `true` or `false`. Do you need `null` somewhere in the code (I didn't find it)?

viniciuskneves commented on src/stories/HdForm.stories.js:
```diff
@@ -72,6 +74,9 @@ const Template = (args, { argTypes }) => ({
       console.log('submitted', params);
     },
+    onFormChange(val) {
+      this.hasFormChanged = val;
```
You can also log it in the actions tab, what do you think?

viniciuskneves reviewed (commented):
Hey @majakomel I think you need to update the PR description since you went for another solution.

majakomel commented:
@viniciuskneves You're so fast! I was testing meanwhile if the changes work fine with the feature I'm implementing in myHomeday before pinging for another review. :)

Thanks for initial suggestions and example, I removed the prop as suggested, it was really not necessary.

viniciuskneves reviewed (approved):
🚀

kuroski commented on src/stories/HdForm.stories.js:
```diff
@@ -72,6 +74,10 @@ const Template = (args, { argTypes }) => ({
     },
+    onFormChange(val) {
+      this.hasFormChanged = val;
+      console.log('hasChanged', val);
```
Don't forget to remove the `console.logs` 😄

majakomel replied:
This was actually suggested by Vini, so we can see logs under actions when trying out the component https://github.com/homeday-de/homeday-blocks/pull/647#discussion_r564651065 :)

kuroski commented on tests/unit/components/form/HdForm.spec.js:
```diff
@@ -138,6 +142,34 @@ describe('HdForm', () => {
+    expect(wrapper.emitted().hasChanged[2][0]).toBe(false);
+  });
+
+  test('a HdForm does not emit hasChanged event if there is no listener', async () => {
```
🚀

kuroski commented on src/components/form/HdForm.vue:
```diff
@@ -31,8 +31,22 @@ export default {
     this.initialFormData = _cloneDeep(this.getFormData());
+
+    if (this.$listeners.hasChanged) {
+      this.$watch('fieldsValues', () => {
```
I don't know if is relevant but maybe we need to make sure to `unwatch` this before the component is destroyed: 

``` javascript
var unwatch = vm.$watch('a', cb)
// later, teardown the watcher
unwatch()
```

I don't know if Vue handles this automatically but maybe it worth testing, otherwise we might face some memory leaks

kuroski reviewed (commented):
Just some comments, but for me everything is great 😄 
Nice work!

ilyasmez reviewed (commented):
Hi @majakomel,
The implementation will do the just just fine!
But I have a few concerns regarding the design of the component's API.

#### Imperative watcher
I see that you add the watcher only if the listener `hasChanged` is set by the parent **when the component is first created**, I assume that you did so to avoid doing unnecessary computations if the parent is not listening to this event, but that will cause problems if the event listener is dynamically added to the component.
To avoid all this I'd suggest to have a normal declarative watcher, and fire the event whenever a change has been made. The computation time we want to save here is really minor.


#### Name and payload
When I read the event name `hasChange` I'm expecting this event to be fired just once (**the first time** the form data changes), but apparently you're firing this event **every time** something changes, so I'd suggest adapting the name, if you want to keep the same logic you can just rename the event to `change` or `data-change`, so it can be listened to using `@change`, which reads as "on change".
You can return the same payload as we are returning when the form is submitted (you can ignore the complicated `isValid` for now) this will keep the component a little bit clean to avoid adding so many flag-like events, and also because `hasChanged` sounds more like a prop instead of an event (did you look into using the `.sync` modifier instead of having an event?)

majakomel commented:
@ilyasmez Thanks for suggestions, I changed the watcher to declarative one and renamed the event to `change` to make it more clear.

As for using the `.sync` modifier on the prop, that was my initial PR https://github.com/homeday-de/homeday-blocks/pull/647/commits/f38ef84b9f2fc6ffa438bbf11725a30a9a874dde - but because it still requires an event to be emitted for prop to be updated, @viniciuskneves suggested to remove the use of prop. I honestly don't have any strong preference for either of solutions.

ilyasmez commented:
> @ilyasmez Thanks for suggestions, I changed the watcher to declarative one and renamed the event to `change` to make it more clear.
> 
> As for using the `.sync` modifier on the prop, that was my initial PR [f38ef84](https://github.com/homeday-de/homeday-blocks/commit/f38ef84b9f2fc6ffa438bbf11725a30a9a874dde) - but because it still requires an event to be emitted for prop to be updated, @viniciuskneves suggested to remove the use of prop. I honestly don't have any strong preference for either of solutions.

Thank you making the changes 🙇 
I would have honestly preferred `.sync` prop if kept the flag-like event, but as we have switched to a generic solution, we won't need it anymore (especially we might end up with many `.sync` props).

I have just one tiny request left, could you please change:
```js
this.$emit('change', hasChanged);
```
To:
```js
this.$emit('change', { hasChanged });
```

The reason is that the second one is more future-proof, so if we want to add any data to the payload later, we won't introduce breaking changes. And it will also make it similar to the on submit payload.

majakomel commented:
@ilyasmez Done :)

ilyasmez reviewed (approved):
🚀 
Nice work!
